app/services/order_cycle_distributed_products.rb

```diff
@@ -0,0 +1,41 @@
+# Finds valid products distributed by a particular distributor in an order cycle
+#
+# If a product without variants is added to an order cycle, and then some
+# variants are added to that product, but not the order cycle, then the master
+# variant should not available for customers to purchase. This class filters
+# out such products so that the customer cannot purchase them.
+class OrderCycleDistributedProducts
+  def initialize(order_cycle, distributor)
+    @order_cycle = order_cycle
+    @distributor = distributor
+  end
+
+  # Returns an ActiveRecord relation without invalid products. Check
+  # #valid_products_distributed_by for details
+  #
+  # @return [ActiveRecord::Relation<Spree::Product>]
+  def relation
+    variants = order_cycle.variants_distributed_by(distributor)
+    products = variants.map(&:product).uniq
```

Comment:
This loads every product individually. I think this would be much more efficient:
```suggestion
    products = Product.where(id: variants.select(:product_id)).uniq
```

That combines it all into one query.
```sql
SELECT DISTINCT "spree_products".*
 FROM "spree_products"
  WHERE "spree_products"."id" IN (SELECT  product_id FROM "spree_variants" ...)
```

Reply:
This inevitably moves the N+1 to the block below, to the variants table, which would then require us to use [includes](https://apidock.com/rails/ActiveRecord/QueryMethods/includes). I think it's wiser to stop here because #3596 might turn this change pointless. We'll see there.